src/Bridges/graph.jl

```diff
@@ -0,0 +1,269 @@
+    empty!(graph.variable_constraint_cost)
+    empty!(graph.variable_dist)
+    empty!(graph.variable_best)
+    graph.variable_last_correct = 0
+    empty!(graph.constraint_edges)
+    empty!(graph.constraint_dist)
+    empty!(graph.constraint_best)
+    graph.constraint_last_correct = 0
+    empty!(graph.objective_edges)
+    empty!(graph.objective_dist)
+    empty!(graph.objective_best)
+    graph.objective_last_correct = 0
+end
+
+function add_edge(graph::Graph, node::VariableNode, edge::Edge)
+    push!(graph.variable_edges[node.index], edge)
+end
+function add_edge(graph::Graph, node::ConstraintNode, edge::Edge)
+    push!(graph.constraint_edges[node.index], edge)
+end
+function add_edge(graph::Graph, node::ObjectiveNode, edge::ObjectiveEdge)
+    push!(graph.objective_edges[node.index], edge)
+end
+function add_variable_node(graph::Graph)
+    push!(graph.variable_edges, Edge[])
+    # Use an invalid index so that the code errors instead return something
+    # incorrect in case `set_variable_constraint_node` is not called.
+    push!(graph.variable_constraint_node, ConstraintNode(-1))
+    push!(graph.variable_constraint_cost, 0)
+    push!(graph.variable_dist, INFINITY)
+    push!(graph.variable_best, 0)
+    return VariableNode(length(graph.variable_best))
+end
+function set_variable_constraint_node(graph::Graph, variable_node::VariableNode,
+                                      constraint_node::ConstraintNode, cost::Int)
+    graph.variable_constraint_node[variable_node.index] = constraint_node
+    graph.variable_constraint_cost[variable_node.index] = cost
+end
+function add_constraint_node(graph::Graph)
+    push!(graph.constraint_edges, Edge[])
+    push!(graph.constraint_dist, INFINITY)
+    push!(graph.constraint_best, 0)
+    return ConstraintNode(length(graph.constraint_best))
+end
+function add_objective_node(graph::Graph)
+    push!(graph.objective_edges, ObjectiveEdge[])
+    push!(graph.objective_dist, INFINITY)
+    push!(graph.objective_best, 0)
+    return ObjectiveNode(length(graph.objective_best))
+end
+
+function bridge_index(graph::Graph, node::VariableNode)
+    bellman_ford!(graph)
+    return graph.variable_best[node.index]
+end
+function bridge_index(graph::Graph, node::ConstraintNode)
+    bellman_ford!(graph)
+    return graph.constraint_best[node.index]
+end
+function bridge_index(graph::Graph, node::ObjectiveNode)
+    bellman_ford!(graph)
+    return graph.objective_best[node.index]
+end
+
+# Update `b.variable_dist`, `b.constraint_dist` `b.variable_best` and
+# `b.constraint_best` for constrained variable types in `variables` and
+# constraint types in `constraints`.
+function updated_dist(graph::Graph, current, edges::Vector{<:AbstractEdge})
+    bridge_index = 0
+    for edge in edges
+        if supports_added_no_update(graph, edge)
+            # `added_dist != INFINITY` because `supports_added_no_update`
+            # returned `true`.
+            dist = 1 + added_dist(graph, edge)
```

Comment:
So, future-proofing: if we want to add weights to the bridges, we should replace the `xxx_dist::Vector{Int}` with `xxx_dist::Vector{Float64}`, then define `weight(edge) = 1.0` and replace `dest = weight(edge) + added_dist(graph, edge)`?

Then it's a matter of solvers being able to define specific edge weights -- not sure how that would compose with different layers?

Reply:
Yes exactly. If the weight is a constraint attribute, that might compose well

Reply:
> If the weight is a constraint attribute, that might compose well

💯 

There's something nice about `MOI.Bridges` defining a module-specific constraint attribute. We probably also need a `AbstractModelAttribute` for the cost of adding a variable.